### PR TITLE
test(client-services): text replication

### PR DESCRIPTION
**Pull request by wittjosiah**

### CODE REVIEW

**Comment by wittjosiah** [comment]
Current dependencies on/for this PR:
* main
  * **PR #2718** <a href="https://app.graphite.dev/github/pr/dxos/dxos/2718" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dxos/dxos/2718?utm_source=stack-comment).

File: packages/sdk/client-services/src/packlets/testing/test-builder.ts [hunk 1]
```diff
@@ -138,3 +140,41 @@ export const syncItems = async (db1: DatabaseBackendProxy, db2: DatabaseBackendP
...
   await testSpace(db2, db1);
 };
+
+export const joinCommonSpace = async ([initialPeer, ...peers]: Client[], spaceKey?: PublicKey): Promise<PublicKey> => {
+  const rootSpace = spaceKey ? initialPeer.echo.getSpace(spaceKey) : await initialPeer.echo.createSpace();
+  assert(rootSpace, 'Space not found.');
+
+  await Promise.all(
+    peers.map(async (peer) => {
+      const success1 = new Trigger<Invitation>();
```
Comment (richburdon) [inline packages/sdk/client-services/src/packlets/testing/test-builder.ts:150]: consider:
success1, success2 => hostDone, guestDone
observable1, observable2 => hostObservable, etc

File: packages/sdk/client-services/src/packlets/tests/text.test.ts [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+      const space = peer.echo.getSpace(real.spaceKey);
+      if (space) {
+        if (!model.peers.has(peer.halo.identity!.identityKey)) {
+          throw new Error(`Expected peer to not be in space: ${peer.halo.identity!.identityKey.truncate()}`);
+        }
+
+        const [document] = space.db.query((obj) => !!obj.content).objects;
+        const text = (document.content.doc as Doc).getText('utf8');
+        if (text.toString() !== model.text) {
+          throw new Error(`Expected text to be "${model.text}" but was "${text}."`);
```
Comment (richburdon) [inline packages/sdk/client-services/src/packlets/tests/text.test.ts:39]: consider:
new Error(`Mismatch: ${JSON.stringify({ expected, actual })}`)
easier to read/diff, and more tolerant of longer text.

File: packages/sdk/client-services/src/packlets/tests/text.test.ts [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+
+  toString() {
+    const peer = this.peer.halo.identity?.identityKey.truncate();
+    return `RemoveText(peer=${peer}, index=${this.index}, length=${this.length})`;
+  }
+}
+
+describe.only('Client text replication', () => {
+  test('property-based tests', async () => {
+    const testBuilder = new TestBuilder();
```
Comment (richburdon) [inline packages/sdk/client-services/src/packlets/tests/text.test.ts:140]: :)

File: packages/sdk/client-services/src/packlets/tests/text.test.ts [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+          model.peers.add(client.halo.identity!.identityKey);
+          real.peers.set(client.halo.identity!.identityKey, client);
+
+          return { model, real };
+        };
+
+        await fc.asyncModelRun(setup, commands);
+      }),
+      {
+        examples: [
```
Comment (richburdon) [inline packages/sdk/client-services/src/packlets/tests/text.test.ts:198]: can you explain?
is this a starting point for more randomized tests?

File: packages/sdk/client-services/src/packlets/tests/text.test.ts [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+  peers: ComplexSet<PublicKey>;
+};
+
+type Real = {
+  spaceKey: PublicKey;
+  peers: ComplexMap<PublicKey, Client>;
+};
+
+const assertState = async (model: Model, real: Real) => {
+  await waitForExpect(() => {
```
Comment (richburdon) [inline packages/sdk/client-services/src/packlets/tests/text.test.ts:28]: is it possible to make this event driven? e.g., test based on throttled update/join events?
i don't immediately see how, but waitFor always makes me uneasy.

File: packages/sdk/client-services/src/packlets/tests/text.test.ts (hunk 4 above)
Comment (wittjosiah) [reply packages/sdk/client-services/src/packlets/tests/text.test.ts:198]: yeah, exactly, I'm expecting we'll fill out more examples once it's working but just trying to get a very simple one passing first to make sure it's structured right as I haven't used this before and am just modelling off of @dmaretskyi 's network manager tests

File: packages/sdk/client-services/src/packlets/tests/text.test.ts (hunk 5 above)
Comment (wittjosiah) [reply packages/sdk/client-services/src/packlets/tests/text.test.ts:28]: this is copied from the network manager property tests, unsure if it's needed yet, we'll see. Wasn't sure if we'd have to wait for replication or not and how easy that would be to do with events in these fuzzy tests

**Pull request merged**